### PR TITLE
Fix redecl chain failures

### PR DESCRIPTION
**Pull request by martong**
Fix some redecl chain failures of Classes, ClassTemplates, FunctionTemplateSpecs.

### CODE REVIEW

File: lib/AST/ASTImporter.cpp [hunk 1]
```diff
@@ -2628,6 +2609,16 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
   const FunctionDecl *FoundByLookup = nullptr;
   FunctionTemplateDecl *FromFT = D->getDescribedFunctionTemplate();
 
+  // Check if we have found an existing definition.  Returns with that
+  // definition if yes, otherwise returns null.
+  auto IsDefinitionFound = [this](
+      FunctionDecl *D, FunctionDecl *FoundFunction) -> Decl * {
```
Comment (balazske) [inline lib/AST/ASTImporter.cpp:2615]: `FindAndMapDefinition`?
`D` can be captured in the lambda, or call the parameter `FromFunction` instead.

File: unittests/AST/ASTImporterGenericRedeclTest.cpp [hunk 2]
```diff
@@ -99,6 +99,46 @@ struct RedeclChain : ASTImporterOptionSpecificTestBase {
...
+        auto *TemplateD = FirstDeclMatcher<FunctionTemplateDecl>().match(
+            ToTU, functionTemplateDecl());
+        auto *FirstSpecD = *(TemplateD->spec_begin());
+        EXPECT_EQ(FirstSpecD->getCanonicalDecl(), To0F->getCanonicalDecl());
+        return;
+      }
+    }
+
+    // The rest: Classes, Functions, etc.
+    EXPECT_EQ(To1->getPreviousDecl(), To0);
```
Comment (balazske) [inline unittests/AST/ASTImporterGenericRedeclTest.cpp:157]: This check is not needed in the other cases?

File: lib/AST/ASTImporter.cpp (hunk 1 above)
Comment (balazske) [reply lib/AST/ASTImporter.cpp:2615]: Can we call `MapImported` here and not `Imported`, that is called by the Decl import?

Comment (martong) [reply lib/AST/ASTImporter.cpp:2615]: Ok, I like the name you suggest.

File: unittests/AST/ASTImporterGenericRedeclTest.cpp (hunk 2 above)
Comment (martong) [reply unittests/AST/ASTImporterGenericRedeclTest.cpp:157]: Good catch, indeed we must check this in case of templates too.

**Comment by martong** [comment]
Update: rebase from ctu-clang7 and fixed @balazske's comments

**Review by balazske — APPROVED** [review]

**Comment by martong** [comment]
Update: rebased to ctu-clang7

**Pull request merged**